### PR TITLE
Finetune CLIP

### PR DESCRIPTION
**Pull request by zhiqiangdon**
*Issue #, if available:*

*Description of changes:*

1. Support finetuning CLIP for image<-->text matching with preset `image_text_similarity`. Results on Flickr30k are summarized [here](https://quip-amazon.com/fhZ7ApT8a2Us/Finetune-CLIP-with-AutoMM).
2. Add MultiNegativesSoftmaxLoss. Support gathering negative samples across GPUs.
3. Add a customized hit rate metric to compute ranking scores. For now we don't use the [torchmetric's hit rate](https://torchmetrics.readthedocs.io/en/stable/retrieval/hit_rate.html) since the current validation dataloader doesn't support it.
4. Add tests for image<-->text finetuning and semantic search.
5. Change `val_` to `validate_` in dataloader to support using `trainer.validate()`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/matcher.py [hunk 1]
```diff
@@ -1153,36 +1224,70 @@ def _predict(
...
+        )
+
+        return sum(score_1.values()) + sum(score_2.values())
+
     def _evaluate_ranking(
         self,
         qr_relevance: Union[pd.DataFrame, dict, list],
         query_data: Union[pd.DataFrame, dict, list],
         response_data: Union[pd.DataFrame, dict, list],
+        label_column: str,
```
Comment (cheungdaven) [inline multimodal/src/autogluon/multimodal/matcher.py:1262]: is label_column compulsory?

File: multimodal/src/autogluon/multimodal/matcher.py [hunk 2]
```diff
@@ -1461,7 +1590,21 @@ def extract_embedding(
         by the neural network's architecture.
         """
         if signature is None:
-            signature = QUERY
+            if self._query or self._response:
+                if isinstance(data, list):
```
Comment (bryanyzhu) [inline multimodal/src/autogluon/multimodal/matcher.py:1594]: I see there are several assertion statements to check if data is a list. My question is, other than list/dict/df, will other data types be expected? Is it possible that users just give a None, or a str, or a float?

File: multimodal/src/autogluon/multimodal/optimization/losses.py [hunk 3]
```diff
@@ -117,3 +135,140 @@ def __init__(self):
...
+        image features of the current process.
+    text_features
+        text features of the current process.
+    local_loss
+        If False, make sure the features on the current GPU have gradients.
+    gather_with_grad
+        Whether to gather all features with gradients enabled.
+    rank
+        Rank of the current process (it should be a number between 0 and world_size-1).
+    world_size
```
Comment (bryanyzhu) [inline multimodal/src/autogluon/multimodal/optimization/losses.py:158]: Just to confirm, the world size here means number of process, not number of GPUs?

File: multimodal/tests/unittests/test_matcher.py [hunk 4]
```diff
@@ -40,16 +41,66 @@ def verify_matcher_save_load(matcher, df, verify_embedding=True, cls=MultiModalM
...
+            query_column=response_column,
+            response_column=query_column,
+        )
+        socre_2 = matcher.evaluate(
+            data=test_df_with_label,
+            query_data=test_query_image_data,
+            response_data=test_response_text_data,
+            metrics=[metric_name],
+            label_column=test_label_column,
+            # query_signature="response",
```
Comment (bryanyzhu) [inline multimodal/tests/unittests/test_matcher.py:76]: commented lines or TODO?

File: multimodal/src/autogluon/multimodal/utils/matcher.py [hunk 5]
```diff
@@ -332,7 +337,7 @@ def semantic_search(
     response_data: Optional[Union[pd.DataFrame, dict, list]] = None,
     query_embeddings: Optional[torch.Tensor] = None,
     response_embeddings: Optional[torch.Tensor] = None,
-    query_chunk_size: int = 100,
+    query_chunk_size: int = 128,
     response_chunk_size: int = 500000,
```
Comment (cheungdaven) [inline multimodal/src/autogluon/multimodal/utils/matcher.py:341]: will 500000 be too large for some machines?

File: multimodal/src/autogluon/multimodal/optimization/utils.py [hunk 6]
```diff
@@ -147,6 +150,79 @@ def compute(self) -> torch.Tensor:
         return f1_score
 
 
+class CustomHitRate(torchmetrics.Metric):
```
Comment (cheungdaven) [inline multimodal/src/autogluon/multimodal/optimization/utils.py:153]: should not this be in the metric.py?

File: multimodal/src/autogluon/multimodal/optimization/losses.py (hunk 3 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/optimization/losses.py:158]: In GPU machines, they are usually the same. In TPU servers, I think it means TPU cores.

File: multimodal/src/autogluon/multimodal/utils/matcher.py (hunk 5 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/utils/matcher.py:341]: sentence transformer (https://www.sbert.net/examples/applications/semantic-search/README.html#sentence_transformers.util.semantic_search) uses this number.

File: multimodal/src/autogluon/multimodal/matcher.py (hunk 1 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/matcher.py:1262]: I think so. We need to know the label column in `qr_relevance`. For datasets without `label_column` in training, we still need it in evaluation.

File: multimodal/tests/unittests/test_matcher.py (hunk 4 above)
Comment (zhiqiangdon) [reply multimodal/tests/unittests/test_matcher.py:76]: Commented lines. Removed.

**Review by sxjscience — APPROVED** [review]
LGTM

File: multimodal/src/autogluon/multimodal/matcher.py (hunk 2 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/matcher.py:1594]: In matching, we currently don't assume other data types. A None, str, or float would encounter error when being converted to dataframe.

Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/matcher.py:1594]: @zhiqiangdon we should really use DataFrame format. Let's revise it in the next PR. It's more urgent to get the API aligned for now.

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/matcher.py:1594]: @sxjscience The `extract_embedding` in matcher should support passing a list, dict, or a dataframe. This is aligned with the predictor.

**Review by bryanyzhu — APPROVED** [review]
LGTM.

**Review by cheungdaven — APPROVED** [review]
